src/esp32/single_encoder.rs

```diff
@@ -0,0 +1,341 @@
+        if status & pcnt_evt_h_lim != 0 {
+            arg.acc.fetch_add(1, Ordering::SeqCst);
+        }
+        if status & pcnt_evt_l_lim != 0 {
+            arg.acc.fetch_add(1, Ordering::SeqCst);
+        }
+    }
+
+    #[inline(always)]
+    #[link_section = ".iram1.pcnt_srv"]
+    unsafe extern "C" fn irq_handler_decrement(arg: *mut core::ffi::c_void) {
+        let arg: &mut PulseStorage = &mut *(arg as *mut _);
+        let mut status = 0;
+        esp_idf_sys::pcnt_get_event_status(arg.unit, &mut status as *mut c_ulong);
+        if status & pcnt_evt_h_lim != 0 {
+            arg.acc.fetch_sub(1, Ordering::SeqCst);
+        }
+        if status & pcnt_evt_l_lim != 0 {
+            arg.acc.fetch_sub(1, Ordering::SeqCst);
+        }
+    }
+}
+
+impl Encoder for Esp32SingleEncoder {
+    fn get_properties(&mut self) -> EncoderSupportedRepresentations {
+        EncoderSupportedRepresentations {
+            ticks_count_supported: true,
+            angle_degrees_supported: false,
+        }
+    }
+    fn get_position(&self, position_type: EncoderPositionType) -> anyhow::Result<EncoderPosition> {
+        match position_type {
+            EncoderPositionType::TICKS | EncoderPositionType::UNSPECIFIED => {
+                let count = self.get_counter_value()?;
+                Ok(EncoderPositionType::TICKS.wrap_value(count as f32))
+            }
+            EncoderPositionType::DEGREES => {
+                anyhow::bail!("Esp32SingleEncoder does not support returning angular position")
+            }
+        }
+    }
+    fn reset_position(&mut self) -> anyhow::Result<()> {
+        self.reset()
+    }
+}
+
+impl SingleEncoder for Esp32SingleEncoder {
+    fn get_direction(&self) -> anyhow::Result<Direction> {
+        Ok(self.dir)
+    }
+    fn set_direction(&mut self, dir: Direction) -> anyhow::Result<()> {
+        let mut reconfigure = false;
+        match dir {
+            Direction::Forwards | Direction::StoppedForwards => {
+                if !self.dir.is_forwards() {
+                    self.config.neg_mode = pcnt_count_inc;
+                    self.config.pos_mode = pcnt_count_inc;
+                    reconfigure = true;
+                }
+            }
+            Direction::Backwards | Direction::StoppedBackwards => {
+                if self.dir.is_forwards() {
+                    self.config.neg_mode = pcnt_count_dec;
+                    self.config.pos_mode = pcnt_count_dec;
+                    reconfigure = true;
+                }
+            }
+        };
+        self.dir = dir;
+        let isr_is_installed = isr_installed();
+        if reconfigure && isr_is_installed {
+            unsafe {
+                match esp_idf_sys::pcnt_counter_pause(self.config.unit) {
+                    ESP_OK => {}
+                    err => return Err(EspError::from(err).unwrap().into()),
+                }
+
+                match esp_idf_sys::pcnt_unit_config(&self.config as *const pcnt_config_t) {
```

Comment:
do we need to go through all of the config steps or can we just change the interrupt function then resume

Reply:
Some of these steps are necessary, but perhaps not all. Reconfiguring the unit disables the high and low events, so those need to be enabled. I found that I had to re-add the glitch filter, but this might actually be an indicator of a bug in the original `setup_pcnt` function. I'll play around with it on hardware tomorrow

Reply:
but does the config changes? or just the interrupts?

Reply:
update: I did have to re-add the glitch filter